compiler: fix make([]T, ...) with big integers on 32-bit systems or less

aykevl opened the pull request:
Previously, this would have resulted in a LLVM verification error because runtime.sliceBoundsCheckMake would not accept 64-bit integers on these platforms.

TODO before merging: add tests.

aykevl commented:
@kyegupov does this fix the bug described in #120?

kyegupov commented on compiler/compiler.go:
```diff
@@ -2182,21 +2182,31 @@ func (c *Compiler) parseExpr(frame *Frame, expr ssa.Value) (llvm.Value, error) {
-			if sliceLen.Type().IntTypeWidth() < c.uintptrType.IntTypeWidth() {
+			checkFunc := "sliceBoundsCheckMake"
+			biggestInt := c.uintptrType
+			biggestIntWidth := biggestInt.IntTypeWidth()
```
the proper name would be "pointerWidth", uintptr is not a "biggest int"

aykevl replied:
No, because it can be wider than a pointer.

kyegupov replied:
Umm, `c.uintptrType` is supposed to be as wide as a pointer; but it is definitely not the "biggest int".

aykevl replied:
Yes, `c.uintptrType` is the same thing as `uintptr` and thus as big as a pointer. But a few lines below it is set to `uint64`, which certainly can be bigger than a pointer:
https://github.com/aykevl/tinygo/pull/129/files/4992faa0e881d0ea40d6cf8abf81d3f96e6344e3#diff-aa244bd5dbd5c4c8e88f1ad97ce253ebR2192
So `pointerWidth` is certainly wrong. Let me know if you have an idea for a better variable name.

deadprogram replied:
Those names `capacityType` and `capacityTypeWidth` seem like a pretty good change to me.

aykevl replied:
Not entirely happy with the other name but it's okay, change made to avoid more bikeshedding.

kyegupov commented on src/runtime/panic.go:
```diff
@@ -62,3 +62,10 @@ func sliceBoundsCheckMake(length, capacity uint) {
 }
+
+// Check for bounds in *ssa.MakeSlice. Supports 64-bit indexes.
+func sliceBoundsCheckMake64(length, capacity uint64) {
```
a related bug:  sliceBoundsCheckMake params are uint, but it's not guaranteed that uint has the same size as uintptr

aykevl replied:
You are right. It looks like I missed this function in ee7c276493547f1e8ea0ded3f91310ed38d0e1ac (but the bug is older than that).

aykevl replied:
On second thought, no. The length and capacity parameters are _not_ the same thing as what goes into the slice but what is supplied by the user. So it has to handle anything that can go into `make([]T, ...)`, including 64-bit pointers on 32-bit platforms for example.

kyegupov replied:
No, my point is, imagine uintptr is 16 bit, int/uint is 32 (Arduino?), len/cap are 16 bit. `biggestInt` will be `uintptr`, but `biggestIntWidth` will be 4. The check `sliceLen.Type().IntTypeWidth() < biggestIntWidth` does fire (2<4), and you "extend" sliceLen to `biggestInt` which is still 16 bit.

And then you call  `sliceBoundsCheckMake` with 16-bit sliceLen/sliceCap, but it expects 32-bit parameters.

aykevl replied:
`c.uintptrType.IntTypeWidth()` is 16 (number of bits) on an Arduino, I just checked.
The `IntTypeWidth` method returns the number of bits the method takes up. The name is a bit confusing, but it does call a method called [`getBitWidth`](https://llvm.org/doxygen/ValueTracking_8cpp.html#a2cb59ea8f9e8543986d40c48acfd24a3) which is more clear in what it returns.

kyegupov commented on compiler/compiler.go:
```diff
@@ -2182,21 +2182,31 @@ func (c *Compiler) parseExpr(frame *Frame, expr ssa.Value) (llvm.Value, error) {
+			if sliceLen.Type().IntTypeWidth() > biggestIntWidth || sliceCap.Type().IntTypeWidth() > biggestIntWidth {
+				// System that is less than 64bit, meaning that the slice make
+				// params are bigger than uintptr.
+				checkFunc = "sliceBoundsCheckMake64"
```
but you are still truncating the capacity below on line 2215

aykevl replied:
Good point, this is an issue, especially when the slice contains elements bigger than a byte. I'll have to think about a proper fix for this.

aykevl replied:
For reference, this is where the main Go compiler checks slice bounds:
https://github.com/golang/go/blob/master/src/runtime/slice.go#L34

kyegupov commented:
@aykevl to verify, you should merge into this https://github.com/aykevl/tinygo/pull/128 (and probably add tests for other int sizes)

aykevl commented:
I've added a few more checks to check whether the multiplication by the element size and truncation is actually safe.

deadprogram commented:
This branch needs to be brought up to date with `dev` branch.

deadprogram commented:
@aykevl any response to the last feedback from @kyegupov on this PR?

deadprogram commented on compiler/compiler.go:
```diff
@@ -2179,29 +2186,50 @@ func (c *Compiler) parseExpr(frame *Frame, expr ssa.Value) (llvm.Value, error) {
 				}
 			}
-			c.createRuntimeCall("sliceBoundsCheckMake", []llvm.Value{sliceLen, sliceCap}, "")
+			// Note: the max element size needs to be doubled to make sure it
```
@aykevl what do you think?

deadprogram commented:
This PR still needs additional work and tests written according to @aykevl so still pending those changes.

aykevl commented:
I think I've fixed everything mentioned in comments.
I've also added tests, which should be equivalent to #128 but more thoroughly (testing all integer types). While making the changes, I confirmed that the `make([]T, ...)` tests are actually run, before that they didn't actually produce a [`*ssa.MakeSlice`](https://godoc.org/golang.org/x/tools/go/ssa#MakeSlice) instruction.
The slice tests compile for Arduino and the part that tests making slices also appears to work (no panics), however there are other errors (probably bugs in the AVR backend) so I can't fully test it.

aykevl commented:
Note: this PR is ready for review.

deadprogram commented:
Looks good now, and working for me. Now merging.
